src/Client.php

```diff
@@ -20,91 +20,107 @@ class Client
+
     public function __construct(
-        $host,
-        HttpClientInterface $httpClient,
-        $apiVersion = 21
+        string $host,
+        AuthenticationInterface $authentication,
+        ClientInterface $httpClient,
+        RequestFactoryInterface $requestFactory,
+        StreamFactoryInterface $streamFactory,
+        int $apiVersion = 36
     ) {
         $this->baseUrl = rtrim($host, '/') . '/api/' . $apiVersion . '/';
         $this->httpClient = $httpClient;
+        $this->requestFactory = $requestFactory;
+        $this->authentication = $authentication;
+        $this->streamFactory = $streamFactory;
     }
 
     /**
-     * @param string $method
-     * @param string $function
-     * @param array  $parameters
-     *
-     * @return ResponseInterface
+     * @throws ClientExceptionInterface
+     * @throws InvalidArgumentException
      * @throws Exception
      */
-    public function request($method, $function, array $parameters = [])
+    public function request(string $method, string $function, array $parameters = []): ResponseInterface
     {
         try {
-            $response = $this->httpClient->request(
-                $method,
-                $this->baseUrl . rtrim($function, '/'),
-                $parameters
-            );
+            $request = $this->requestFactory
+                ->createRequest($method, $this->baseUrl . rtrim($function, '/'))
+                ->withHeader('Accept', 'application/json');
+
+            if (!empty($parameters)) {
+                $body = json_encode($parameters);
+                if (false === $body) {
+                    throw new InvalidArgumentException(
+                        'Request parameters could not be converted to json: ' . json_last_error_msg()
+                    );
+                }
+
+                $request = $request
+                    ->withBody($this->streamFactory->createStream($body))
+                    ->withHeader('Content-Type', 'application/json');
+            }
+
+            $request = $this->authentication->authenticate($request);
+            return $this->httpClient->sendRequest($request);
         } catch (Exception $exception) {
-            if (method_exists($exception, 'getResponse')) {
+            if ($exception instanceof ExceptionResponseInterface && null !== $exception->getResponse()) {
                 return $exception->getResponse();
             }
 
             throw $exception;
         }
-
-        return $response;
     }
 
     /**
-     * @param string $function
-     * @param array  $parameters
-     *
-     * @return ResponseInterface
+     * @throws ClientExceptionInterface
+     * @throws InvalidArgumentException
      * @throws Exception
      */
-    public function get($function, array $parameters = [])
+    public function get(string $function, array $parameters = []): ResponseInterface
     {
         $uri = $function;
-        if ([] !== $parameters) {
+        if (!empty($parameters)) {
```

Comment:
it is faster to compare array with an an empty one
```suggestion
        if ([] !== $parameters) {
```